### PR TITLE
bugfix: deferred tasks does not cancel when DAG is marked fail

### LINKED ISSUE #20580: Triggers are not terminated when DAG is mark failed
### Apache Airflow version

2.2.3 (latest released)

### What happened

Hello, I quite excited to new [Deferrable ("Async") Operators in AIP-40 ](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=177050929), and going to adapt this by develop new async version of ExternalTaskSensor. But while doing, I find out that Deferrable ("Async") Operators are not canceled when DAG is mark failded/success.
Screenshot from 2021-12-30 14-07-44
When I mark DAG as failed, `wait_task` is canceled (changed to `failed` state), but `wait_task_async` still in `deferred` state and triggerer is keep poking.
Screenshot from 2021-12-30 14-08-06
image

### What you expected to happen

Deferrable ("Async") Operators should be canceled as sync version of operators

### How to reproduce

Testing DAG.
```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.operators.bash import BashOperator
from airflow.sensors.external_task import ExternalTaskSensor, ExternalTaskSensorAsync

with DAG(
    'tutorial_async_sensor',
    default_args={
        'depends_on_past': False,
        'email': ['airflow@example.com'],
        'email_on_failure': False,
        'email_on_retry': False,
        'retries': 1,
        'retry_delay': timedelta(minutes=5),
    },
    description='A simple tutorial DAG using external sensor async',
    schedule_interval=timedelta(days=1),
    start_date=datetime(2021, 1, 1),
    catchup=False,
    tags=['example'],
) as dag:
    t1 = ExternalTaskSensorAsync(
        task_id='wait_task_async',
        external_dag_id="tutorial",
        external_task_id="sleep",
        execution_delta=timedelta(hours=1),
        poke_interval=5.0
    )

    t2 = ExternalTaskSensor(
        task_id='wait_task',
        external_dag_id="tutorial",
        external_task_id="sleep",
        execution_delta=timedelta(hours=1),
        poke_interval=5.0
    )

    t3 = BashOperator(
        task_id='echo',
        depends_on_past=False,
        bash_command='echo Hello world',
        retries=3,
    )
    [t1, t2] >> t3
```

for `ExternalTaskSensorAsync`, see #20583

### Operating System

Ubuntu 20.04

### Versions of Apache Airflow Providers

_No response_

### Deployment

Virtualenv installation

### Deployment details

run `airflow standalone`

### Anything else

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dungdm93**
bugfix: deferred tasks does not cancel when DAG is marked fail #20580 and fix typing
closes: #20580

### CODE REVIEW

File: airflow/api/common/experimental/mark_tasks.py [hunk 1]
```diff
@@ -335,18 +336,18 @@ def set_dag_run_state_to_failed(dag, execution_date, commit=False, session=None)
...
-    # Mark only RUNNING task instances.
+    # Mark only running task instances.
     task_ids = [task.task_id for task in dag.tasks]
     tis = (
         session.query(TaskInstance)
         .filter(
             TaskInstance.dag_id == dag.dag_id,
             TaskInstance.execution_date == execution_date,
             TaskInstance.task_id.in_(task_ids),
+            TaskInstance.state.in_(State.running),
```
Comment (uranusjr) [inline airflow/api/common/experimental/mark_tasks.py:349]:
```suggestion
            TaskInstance.state == TaskInstanceState.running,
```

File: airflow/api/common/experimental/mark_tasks.py [hunk 2]
```diff
@@ -70,7 +71,7 @@ def set_state(
     past: bool = False,
     state: TaskInstanceState = TaskInstanceState.SUCCESS,
     commit: bool = False,
-    session=None,
+    session: SASession = None,
```
Comment (uranusjr) [inline airflow/api/common/experimental/mark_tasks.py]: Please search for `session: Session = NEW_SESSION` for a more idomatic way to annotate this.

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/api/common/experimental/mark_tasks.py [hunk 3]
```diff
@@ -335,18 +336,18 @@ def set_dag_run_state_to_failed(dag, execution_date, commit=False, session=None)
...
     tis = (
         session.query(TaskInstance)
         .filter(
             TaskInstance.dag_id == dag.dag_id,
             TaskInstance.execution_date == execution_date,
             TaskInstance.task_id.in_(task_ids),
+            TaskInstance.state.in_(State.running),
         )
-        .filter(TaskInstance.state == State.RUNNING)
     )
```
Comment (uranusjr) [inline airflow/api/common/experimental/mark_tasks.py]: (Note for other reviewers) From what I can tell, this is the only functional change; all others are Mypy fixes.

Comment (dungdm93) [reply airflow/api/common/experimental/mark_tasks.py]: Yes. That's right.

File: airflow/api/common/experimental/mark_tasks.py (hunk 2 above)
Comment (dungdm93) [reply airflow/api/common/experimental/mark_tasks.py]: Thanks

**Review by uranusjr — APPROVED** [review]

**Review by ephraimbuddy — APPROVED** [review]

**Pull request merged**